### PR TITLE
fix: wrong output shape when batch_size is None in NewtonRaphson

### PR DESCRIPTION
**Pull request by tanguy-marchand**
## Summary
The current implementation of NewtonRaphson is buggy when batch_size = None. Indeed, in this case, the batch_size is set to n_sample. However, in the predict method, a dataloader was called with batch_size=None, which make the dataloader drop the batch dimension, and set a batch size of 1.

Therefore, the shape of the prediction was missing one dimension. For exemple, setting BATCH_SIZE=None in the test `test_pytorch_nr_algo_performance` would result in a fail, as the MAE metric was not accurate in that situation.

Simply erasing the over-writing of the _local_predict solves that issue. Indeed, the parent _local_predict is checking at the IndexGenerator batch_size which is set to n_samples if batch_size = None. There is no need to overwrite it.

## Notes

- Adding a specific test
- Fixing the bug

## Please check if the PR fulfills these requirements

- [ X] If the feature has an impact on the user experience, the [changelog](https://github.com/substra/substrafl/blob/main/CHANGELOG.md) has been updated
- [ X] Tests for the changes have been added (for bug fixes / features)
- [ X] Docs have been added / updated (for bug fixes / features)
- [ X] The commit message follows the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification

### CODE REVIEW

File: substrafl/algorithms/pytorch/torch_newton_raphson_algo.py [hunk 1]
```diff
@@ -249,30 +249,6 @@ def _local_train(
...
-        predictions = torch.Tensor([])
-        with torch.inference_mode():
-            for x in predict_loader:
-                x = x.to(self._device)
-                predictions = torch.cat((predictions, self._model(x)), 0)
-
-        predictions = predictions.cpu().detach()
-
-        self._save_predictions(predictions, predictions_path)
-
```
Comment (ThibaultFy) [inline substrafl/algorithms/pytorch/torch_newton_raphson_algo.py:275]: You cannot delete the overwrite of the local predict because, in the NR algorithm, the index generator is only defined for train task, and not for predict task. As the index generator will be set to None, it'll raise an error BatchSizeNotFoundError (cf [here](https://github.com/Substra/substrafl/blob/918c900c14c02cecb51f79551cd39262123c1fae/substrafl/algorithms/pytorch/torch_base_algo.py#L148))

The "slow" tests are not run on each PR. To see this impact you'll need to run `pytest tests/algorithms/pytorch/test_newton_raphson.py` locally.

File: substrafl/algorithms/pytorch/torch_newton_raphson_algo.py [hunk 2]
```diff
@@ -249,30 +249,6 @@ def _local_train(
...
-        """Executes the following operations:
-
-            * Create the torch dataloader using the batch size given at the ``__init__`` of the class
-            * Sets the model to `eval` mode
-            * Returns the predictions
-
-        Args:
-            predict_dataset (torch.utils.data.Dataset): predict_dataset build from the x returned by the opener.
-        """
-        predict_loader = torch.utils.data.DataLoader(predict_dataset, batch_size=self._batch_size)
```
Comment (ThibaultFy) [inline substrafl/algorithms/pytorch/torch_newton_raphson_algo.py:262]: What you can do instead to remove the bug, is to write `predict_loader = torch.utils.data.DataLoader(predict_dataset, batch_size=self._batch_size or len(predict_dataset))`

File: tests/algorithms/pytorch/test_newton_raphson.py [hunk 3]
```diff
@@ -206,6 +210,27 @@ def forward(self, x):
...
+    [(5, 3), (5, 1), (10, 2), (1, 1), (1, 10)],
+)
+def test_prediction_shape(session_dir, perceptron, torch_algo, n_samples, y_shape):
+    x_shape = 15
+
+    x_train = np.zeros([n_samples, x_shape])
+    y_train = np.ones([n_samples, y_shape])
+
+    model = perceptron(linear_n_col=x_shape, linear_n_target=y_shape)
+    my_algo = torch_algo(model=model, batch_size=1)
```
Comment (ThibaultFy) [inline tests/algorithms/pytorch/test_newton_raphson.py:224]: What you want to test is the behaviour if batch_size is `None`. Why set it to `1` ?

File: tests/algorithms/pytorch/test_newton_raphson.py [hunk 4]
```diff
@@ -206,6 +210,27 @@ def forward(self, x):
...
+
+    my_algo.train(datasamples=(x_train, y_train), _skip=True)
+    prediction_folder = Path(tempfile.mkdtemp(dir=session_dir))
```
Comment (ThibaultFy) [inline tests/algorithms/pytorch/test_newton_raphson.py:227]: `session_dir` is already a temp folder. You can use `prediction_path = session_dir / "NR_predictions"`

File: tests/algorithms/pytorch/test_newton_raphson.py [hunk 5]
```diff
@@ -206,6 +210,27 @@ def forward(self, x):
...
+def test_prediction_shape(session_dir, perceptron, torch_algo, n_samples, y_shape):
+    x_shape = 15
+
+    x_train = np.zeros([n_samples, x_shape])
+    y_train = np.ones([n_samples, y_shape])
+
+    model = perceptron(linear_n_col=x_shape, linear_n_target=y_shape)
+    my_algo = torch_algo(model=model, batch_size=1)
+
+    my_algo.train(datasamples=(x_train, y_train), _skip=True)
```
Comment (ThibaultFy) [inline tests/algorithms/pytorch/test_newton_raphson.py:226]: You don't want to call the `train` method here.

File: tests/algorithms/pytorch/test_newton_raphson.py [hunk 6]
```diff
@@ -206,6 +210,27 @@ def forward(self, x):
...
+    prediction_folder = Path(tempfile.mkdtemp(dir=session_dir))
+    my_algo.predict(datasamples=(x_train, y_train), predictions_path=prediction_folder, _skip=True)
+    file_name = str(prediction_folder).split("/")[-1]
+    prediction = np.load(os.path.join(session_dir, file_name, file_name + ".npy"))
```
Comment (ThibaultFy) [inline tests/algorithms/pytorch/test_newton_raphson.py:230]: With the change on `prediction_path`, you should just do `np.load(prediction_path)`

File: tests/algorithms/pytorch/test_newton_raphson.py (hunk 3 above)
Comment (ThibaultFy) [reply tests/algorithms/pytorch/test_newton_raphson.py:224]: You might want to parametrise it ?

File: tests/algorithms/pytorch/test_newton_raphson.py [hunk 7]
```diff
@@ -206,6 +210,27 @@ def forward(self, x):
         my_algo.train(datasamples=(x_train, y_train), _skip=True)
 
 
+@pytest.mark.parametrize(
+    "n_samples,y_shape",
+    [(5, 3), (5, 1), (10, 2), (1, 1), (1, 10)],
```
Comment (ThibaultFy) [inline tests/algorithms/pytorch/test_newton_raphson.py:215]: In general, we try to limit the parametrization in order to limit the number of tests to whats really needed :)

**Review by ThibaultFy — CHANGES_REQUESTED** [review]
Thanks a lot for the catch and the PR ! 
As said in the comment, your solution will fail on the slow tests that are not run on every PR due to computational reason. You can try to run them locally to see the impact. Especially, the `test_pytorch_nr_algo_performance` will fail if you remove the _local_predict from NR algo.
I put a lot of comments, to help the PR to match the good practice of the repo, but the spirit is exactly that, impressive... So congrats and thanks a lot !

**Review by ThibaultFy — APPROVED** [review]
Thanks a lot for the PR, that's perfect :D
